dask/bag/core.py

```diff
@@ -244,20 +280,39 @@ def __init__(self, dsk, name, npartitions):
         self.npartitions = npartitions
         self.str = StringAccessor(self)
 
-    def map(self, func):
+    def map(self, func, **kwargs):
         """ Map a function across all elements in collection
 
         >>> import dask.bag as db
         >>> b = db.from_sequence(range(5))
         >>> list(b.map(lambda x: x * 10))  # doctest: +SKIP
         [0, 10, 20, 30, 40]
+
+        Any **kwargs are passed through to func along with the (maybe unpacked)
+        item argument(s).  Values in **kwargs must either be a non-dask python
+        literal or a dask.bag.Item.
+
+        If a dask.Bag.Item is used its computed value is passed to the map
+        function.  This means that whatever computations are implied by any
+        Item values block any of the map work.  Example:
+
+        >>> b = db.from_sequence(range(5))
+        >>> b.map(
+        ...     lambda x, total=0: float(x) / total,
+        ...     total=b.sum()).sum().compute()
+        1.0
```

Comment:
I think we should start with the literal value.  This gets the common case
out first, which I think is as if not more important.

Adding something like b.sum() sounds fine too.  I recommend keeping it
inline but moving the lambda out, perhaps replace it with a function div?
I wouldn't worry about `float(x)`, from __future__ import division should
have you covered.

On Tue, Mar 15, 2016 at 2:58 PM, Joshua T Corbin <notifications@github.com>
wrote:

> In dask/bag/core.py
> <https://github.com/dask/dask/pull/1054#discussion_r56248514>:
>
> >          """ Map a function across all elements in collection
> >
> >          >>> import dask.bag as db
> >          >>> b = db.from_sequence(range(5))
> >          >>> list(b.map(lambda x: x * 10))  # doctest: +SKIP
> >          [0, 10, 20, 30, 40]
> > +
> > +        Any **kwargs are passed through to func along with the (maybe unpacked)
> > +        item argument(s).  Values in **kwargs must either be a non-dask python
> > +        literal or a dask.bag.Item.
> > +
> > +        If a dask.Bag.Item is used its computed value is passed to the map
> > +        function.  This means that whatever computations are implied by any
> > +        Item values block any of the map work.  Example:
> > +
> > +        >>> b = db.from_sequence(range(5))
> > +        >>> b.map(
> > +        ...     lambda x, total=0: float(x) / total,
> > +        ...     total=b.sum()).sum().compute()
> > +        1.0
>
> Sure I can totally see having an example in doc string with a constant
> python value, but we also want one with a dask value as well, I think
> simply using temporary variables to make the code clearer, rather than
> going along with the penchant for heavy inlining that dask's source code
> has would help:
>
> >>> nums = db.from_sequence(...)>>> nums_total = nums.sum()>>> normalized = b.map(lambda x, total=0: float(x) / total, total=nums_total)>>> normalized.sum()1.0
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/dask/dask/pull/1054/files/99fc7e7df80eb1def4471dda1611e5e7a577e248#r56248514>
>